### ISSUE TITLE
[Bug]: Browser console error `net::ERR_OUT_OF_MEMORY` when uploading large files

### ISSUE THREAD

**Issue by blmhemu** [nextcloud/server#43627]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

When upload large files (~7.5 GB) the upload fails after sometime - Exactly from chunk 205 onwards. This behaviour is consistent and repeatable. Please see the additional info section.

### Steps to reproduce

1.Create nextcloud apache docker compose setup.
2.Login to the instance and upload a large file. (I uploaded a raw fedora image)
3.Check both the network tab and the console tab for error logs.

### Expected behavior

Uploads should succeed.

### Installation method

Community Docker image

### Nextcloud Server version

28

### Operating system

Other

### PHP engine version

PHP 8.2

### Web server

Apache (supported)

### Database engine version

PostgreSQL

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
Error seems to be occurring on the client js side.

Typical system. Will be happy to provide more details if you think needed.
```

### List of activated Apps

```shell
Fresh install
```

### Nextcloud Signing status

```shell
NA
```

### Nextcloud Logs

```shell
Checked the logs - nothing relevant.
```

### Additional info

It looks like client js is running out of memory - may be it is storing all the chunks in memory and not freeing them up after upload.

![image](https://github.com/nextcloud/server/assets/19410501/b4bed631-dfde-4181-90ca-b6dca69067af)

FWIW, I followed the large file upload section in nextcould docs as well. It is also reproducible in fpm images with caddy.

**Comment by joshtrichards** [comment]
Possibly related to #42704

**Comment by blmhemu** [comment]
@joshtrichards I see that issue has been closed, is there a docker build i can test and see if i can repro this ?

**Comment by solracsf** [comment]
See https://github.com/szaimen/nextcloud-easy-test/pkgs/container/nextcloud-easy-test

**Comment by blmhemu** [comment]
A related question (may be another bug) - When i execute `php occ config:app:get files max_chunk_size` i get empty response, but when i upload files, i see that they are chunked - also see the same in uploads folder. Is this a known behaviour ?

**Comment by blmhemu** [comment]
@solracsf I could not repro the bug using your image 🚀

**Comment by solracsf** [comment]
Closing as per https://github.com/nextcloud/server/issues/42704#issuecomment-1941950753

**Comment by joshtrichards** [comment]
> When I execute php occ config:app:get files max_chunk_size I get empty response, I expected to see 10MB or so (in bytes), but when i upload files, i see that they are chunked - also see the same in uploads folder. Is this the expected behaviour ?

It'll only return a value if the hard-coded default has been overridden. The default is[10 MiB](https://github.com/nextcloud/server/blob/85f28856cfd75fe17f490f098c5657090bc00412/apps/files/lib/App.php#L63).

> This is a clean setup and from what I understand, chunking is enabled by default.

[Correct](https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/big_file_upload_configuration.html#adjust-chunk-size-on-nextcloud-side).

**Comment by blmhemu** [comment]
+1 Exists on 28.0.3 - @solracsf mind re-opening ?

Proofs Attached:
<img width="673" alt="Screenshot" src="https://github.com/nextcloud/server/assets/19410501/6b5114db-ac30-45e1-987b-3d00ef85a988">

<img width="1485" alt="Screenshot 2024-03-06 at 12 30 29 AM" src="https://github.com/nextcloud/server/assets/19410501/fdf87ff0-8345-489d-8842-f254a57ff2c3">
<img width="1532" alt="Screenshot 2024-03-06 at 12 29 01 AM" src="https://github.com/nextcloud/server/assets/19410501/a35b33c2-7cf7-4c8f-9d7e-b6f9fb2106a1">

**Comment by osscombat** [comment]
> +1 Exists on **28.0.3** - @solracsf mind re-opening ?
> 
> Proofs Attached: <img alt="Screenshot" width="673" src="https://private-user-images.githubusercontent.com/19410501/310248050-6b5114db-ac30-45e1-987b-3d00ef85a988.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0OTc3MjgsIm5iZiI6MTcxMDQ5NzQyOCwicGF0aCI6Ii8xOTQxMDUwMS8zMTAyNDgwNTAtNmI1MTE0ZGItYWMzMC00NWUxLTk4N2ItM2QwMGVmODVhOTg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDEwMTAyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiZjJhOWYyZTk3OTdkZmFhZmY2ZjUzYmFiNGQ4MmI3ZmU3ZDM0ODU3YmJjN2Q2MDM4ZDcxZjY1YjUwM2Y4YjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wMCpnjaQxPxH4E64ptqHGDleXwWKROPv0DBBrvyxRGc">
> 
> <img alt="Screenshot 2024-03-06 at 12 30 29 AM" width="1485" src="https://private-user-images.githubusercontent.com/19410501/310248344-fdf87ff0-8345-489d-8842-f254a57ff2c3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0OTc3MjgsIm5iZiI6MTcxMDQ5NzQyOCwicGF0aCI6Ii8xOTQxMDUwMS8zMTAyNDgzNDQtZmRmODdmZjAtODM0NS00ODlkLTg4NDItZjI1NGE1N2ZmMmMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDEwMTAyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM0YTU3YzM3YjdkZDA0ZTVjMzkwYWEyNTc4Y2JkMmFjOTcyZjI4MTFkNzE5ZWE4ZDdiMWUyY2E1MmMyMDFjYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.UW5vNVcEDb0Anya-K7lzt66LN0n29vTMn3_qOVf_eEs"> <img alt="Screenshot 2024-03-06 at 12 29 01 AM" width="1532" src="https://private-user-images.githubusercontent.com/19410501/310248469-a35b33c2-7cf7-4c8f-9d7e-b6f9fb2106a1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0OTc3MjgsIm5iZiI6MTcxMDQ5NzQyOCwicGF0aCI6Ii8xOTQxMDUwMS8zMTAyNDg0NjktYTM1YjMzYzItN2NmNy00YzhmLTlkN2UtYjZmOWZiMjEwNmExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDEwMTAyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllZjdiNWM2MzMxMDZhMDBlZjZiYTUyMTJlMDEyMWMxZDEzMjMzNzcxYjdlMmNmYmQ0MjlmMGRjYTQ0MjdkN2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Up_FCOzU1Y5ytqa1B47GJdSPv2P3dB0aR_gapKUuTFM">
> Notice that it fails exactly at chuck 205 everytime (even for the above report it seems).

actually it depends on your chunk size settings, you can calculate that 205 * default 10mb chunk size equals to ~2Gb.

**Comment by EuleMitKeule** [comment]
![Screenshot 2024-03-16 023500](https://github.com/nextcloud/server/assets/47421406/e85f6d53-7fa5-494e-9960-4dc3ecf064bb)
Same problem on 23.0.3

**Comment by coinfastman** [comment]
We have the same problem! It arose after the transition from version 27 to 28. I was able to understand this error in more detail and this is what I found out. Our problem only occurs when a reverse proxy is used. If you go directly to the web using an IP address, files of any size are upload without any problems; as soon as the download goes through a reverse proxy, a break can occur at any time. Then we managed to find out that the break and the OUT OF MEMORY error depend on the location on the computer disk from which the file is being uploaded. If the file size is larger than the free space on drive C, then this is a 100 percent guarantee that the upload will fail. For some reason, when working through a reverse proxy, the browser begins to duplicate all chunks on drive C.

**Comment by skjnldsv** [comment]
@osscombat and @blmhemu are you also using a reverse proxy?

**Comment by osscombat** [comment]
> @osscombat and @blmhemu are you also using a reverse proxy?

yep

**Comment by skjnldsv** [comment]
Is it possible that you're dropping some headers with your reverse proxy setup?

**Comment by blmhemu** [comment]
@skjnldsv - I did the tests both with and without reverse proxy (caddy) and found the same results in both cases.

**Comment by skjnldsv** [comment]
Alright! Thanks for the help :)
Are you also using 32bits server and/or client?

**Comment by osscombat** [comment]
> Alright! Thanks for the help :) Are you also using 32bits server and/or client?

I use only 64bit everywhere, browsers as well.

**Comment by blmhemu** [comment]
64 - bit everything here as well.

**Comment by blmhemu** [comment]
Chrome on MacOS (both latest version)

**Comment by osscombat** [comment]
> Chrome or Firefox? What OS?

Windows 10/11 and latest Chrome/Edge, 64bit everything.

**Comment by coinfastman** [comment]
Our problem was observed in Windows 10/11 64bit MacOS latest version. Browsers checked Chrome Firefox, the problem is the same everywhere. Regardless of external systems and browser frameworks, when the laptop’s hard drive and insufficient memory capacity, a break occurs.

**Comment by skjnldsv** [comment]
> ![Screenshot 2024-03-16 023500](https://private-user-images.githubusercontent.com/47421406/313364173-e85f6d53-7fa5-494e-9960-4dc3ecf064bb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIzMTk3NzIsIm5iZiI6MTcxMjMxOTQ3MiwicGF0aCI6Ii80NzQyMTQwNi8zMTMzNjQxNzMtZTg1ZjZkNTMtN2ZhNS00OTRlLTk5NjAtNGRjM2VjZjA2NGJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA1VDEyMTc1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzZTc0OTE2NTE2MTMyMWMxN2U4OGU2NTgzYzA2NjZiNzQ5ODYyODAxYzljM2I1ZDM5MTc1MmMxMzBlOGMyNGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kSnmadyyG9K7A8p3IYQ1you5tc8pxJNi_QzlnAT46vc) Same problem on 23.0.3

Your issue is different. Please check the documentation on "big file upload" and open a new ticket if it still occurs after fixing your config (looks like a webserver timeout)

**Comment by skjnldsv** [comment]
Alright, I tried many things.
If my dev tools are opened, I am indeed able to notice a very huge RAM usage.
But without it, the garbage collection works fine and both FF and Chrome manage to keep the ram usage low.

@blmhemu and @osscombat you both seem to be using the same server, right?
Have any of you tried with a different server (like try.nextcloud.com) and/or with a different browser?
It seems you're the only two facing that issue :thinking:

**Comment by skjnldsv** [comment]
@coinfastman you haven't given much data, are you also experiencing a `net::ERR_OUT_OF_MEMORY` error in your browser? Can you share your console log please (screenshot showing the error)

**Comment by osscombat** [comment]
> Alright, I tried many things. If my dev tools are opened, I am indeed able to notice a very huge RAM usage. But without it, the garbage collection works fine and both FF and Chrome manage to keep the ram usage low.
> 
> @blmhemu and @osscombat you both seem to be using the same server, right? Have any of you tried with a different server (like try.nextcloud.com) and/or with a different browser? It seems you're the only two facing that issue 🤔

I'm using my own NC 28.0.4 instance. The Firefox have the same issue, but the desktop NC client uploads just fine. I think this is just a very rare scenario to upload 2+ Gb files via browser and a reverse proxy, that's why not so many complaints.

**Comment by skjnldsv** [comment]
What's your reverse proxy? Can you give us a bit more feedback so we can try to reproduce the issue?

**Comment by osscombat** [comment]
> What's your reverse proxy? Can you give us a bit more feedback so we can try to reproduce the issue?

I use nginx, pretty standard setup with LE, nothing special, nextcloud.conf:

```
map $http_upgrade $connection_upgrade {
    default upgrade;
    '' close;
}

    server {
    
    server_name SERVER.DOMAIN.COM;
    set $upstream 192.168.1.100:80;


    location / {

    add_header Strict-Transport-Security "max-age=15552000; includeSubdomains; preload;";
    proxy_pass http://$upstream;

    proxy_http_version 1.1;
    proxy_set_header "Connection" ""; 
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";

    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Port $server_port;    
    proxy_set_header X-Forwarded-Host $host;

    client_max_body_size 0;
    client_body_timeout 3600s;

    proxy_max_temp_file_size 0;
    proxy_request_buffering off;
    proxy_buffering off;

# Safari IOS fix
    proxy_cookie_path / /;
    proxy_set_header Cookie $http_cookie;

}
    location ^~ /.well-known {
        location = /.well-known/carddav { return 301 $scheme://$http_host/remote.php/dav/; }
        location = /.well-known/caldav  { return 301 $scheme://$http_host/remote.php/dav/; }
        location = /.well-known/nodeinfo { return 301 $scheme://$http_host/index.php/.well-known/nodeinfo; }
        location = /.well-known/webfinger { return 301 $scheme://$http_host/index.php/.well-known/webfinger; }

        return 301 $scheme://$http_host/index.php$request_uri;
    }

    location /ocm-provider { 
        return 301 $scheme://$host/index.php/ocm-provider;
   }

    listen 443 ssl; # managed by Certbot

    ssl_certificate /etc/letsencrypt/live/SERVER.DOMAIN.COM/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/SERVER.DOMAIN.COM/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot
    
}    server {
    if ($host = SERVER.DOMAIN.COM) {
        return 301 https://$host$request_uri;
    } # managed by Certbot


    server_name SERVER.DOMAIN.COM;
    listen 80;
        return 404; # managed by Certbot

}
```

**Comment by coinfastman** [comment]
@skjnldsv I provide a screenshot and change of nginx reverse proxy. Unfortunately, I need to paint over some elements for safety reasons. 
After the download is interrupted, you can immediately notice that the C:\Users\Alex\AppData\Local\Google\Chrome\User Data\Default\blob_storage directory has grown in size and all frames transferred to the server are now in this directory


<img width="1478" alt="Снимок экрана 2024-04-06 в 11 49 26" src="https://github.com/nextcloud/server/assets/33808421/16d23f66-f6dc-41fa-8829-c57272fb7bad">

`server {
    listen 80;
    server_name ***;
    return 301 https://$host$request_uri;
}

server {
    listen 443 ssl;
    server_name ****;
    access_log /var/log/nginx/cloud.access.log;
    error_log  /var/log/nginx/cloud.error.log;
    ssl_certificate      /etc/nginx/certs/***.crt;
    ssl_certificate_key  /etc/nginx/certs/***.key;
    client_max_body_size 25000M;
    ssl_session_timeout  5m;
    ssl_protocols TLSv1.2;
    ssl_ciphers 'ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256';
    ssl_prefer_server_ciphers on;
    proxy_request_buffering off;
    proxy_max_temp_file_size 0;

    add_header Strict-Transport-Security max-age=31536000;

    location / {
        proxy_pass         http://****:8081;
        proxy_set_header   Accept-Encoding "";
        proxy_set_header   Host $host;
        proxy_set_header   X-Real-IP  $remote_addr;
        proxy_set_header   X-Forwarded-Proto $scheme;
        proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
    }

    location /.well-known/carddav {
        return 301 $scheme://$host/remote.php/dav;
    }
    
    location /.well-known/caldav {
        return 301 $scheme://$host/remote.php/dav;
    }

    location /.well-known/webfinger {
        return 301 $scheme://$host/index.php$uri;
    }

    location /.well-known/nodeinfo {
        return 301 $scheme://$host/index.php$uri;
    }

}
`

**Comment by blmhemu** [comment]
@skjnldsv hey ! This issue, led me to finally create a (long overdue) dev environment - I created a perfectly new instance and the bug is still repeatable. FWIW, I disabled all the extensions and the file I am uploading is `Fedora-Server-39-20231103.n.0.aarch64.raw`

**Comment by osscombat** [comment]
> I can reproduce this issue with Chromium and Firefox (not the error but I see memory consumption increases to ~8GB). I also was able to fix this, for Chromium it is fully resolved for Firefox there is a browser bug: If you open the dev tools the request data is not cleared -> memory is not freed. But this is a known memory issue with (Firefox) devtools.
> 
> I will create the patch soon.

Yes, the issue is resolved with the release 28.0.5, thank you!
